### PR TITLE
Feat/federation

### PR DESCRIPTION
**Pull request by steve-keep**
## Summary by CodeRabbit

- **New Features**
  - Introduced federation tests using GitHub Actions to ensure subgraph compatibility with Apollo Federation.
  - Added a new example app for federation compatibility testing, including detailed setup instructions and integration tests.
  - Introduced new data structures and GraphQL entities for products, users, and inventory within the federation example.

- **Documentation**
  - Added README files with instructions for running federation integration tests and setting up the example app.
  
- **Configuration**
  - Updated `.gitignore` and `.vscode/launch.json` to support federation tests and examples.
  - Added Docker and Docker Compose support for federation integration tests.
  - Introduced TypeScript configuration for the federation example.

### CODE REVIEW

**Comment by steve-keep** [comment]
@coderabbitai review

File: src/packages/core/src/decorators/entity.ts [hunk 1]
```diff
@@ -26,6 +26,8 @@ export function Entity<G = unknown>(
 
 		// Let's make sure the new name is set on the target
 		Object.defineProperty(target, 'name', { value: name });
+		// Lets also set the __typename on the prototype for resolving union types
+		Object.defineProperty(target.prototype, '__typename', { value: name });
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/core/src/decorators/entity.ts:30]: Understood and happy for that risk as `__typename` is the standard here.

File: src/examples/federation/integration/README.md [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+Federation v2 compatibility
+*************
+@link PASS
+@shareable PASS
+@tag PASS
+@override PASS
+@inaccessible PASS
+@composeDirective PASS
+@interfaceObject PASS
+```
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/federation/integration/README.md:52]: This is not actually bash.

File: src/packages/core/src/federation/directives.ts [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+		graphweaverMetadata.collectEntityInformation({
+			...entity,
+			directives: {
+				...(entity.directives ? entity.directives : {}),
+				key: {
+					fields: entity.primaryKeyField ?? 'id',
+				},
+			},
+		});
+	}
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/core/src/federation/directives.ts:45]: In Federation it is possible that some entities do not have an ID field.

File: src/examples/federation/src/backend/schema/product-research.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+import { CaseStudy } from './case-study';
+import { data } from '../data';
+
+class JsonDataProvider extends BaseDataProvider<ProductResearch> {
+	findOne(filter: Filter<ProductResearch>): Promise<ProductResearch> {
+		const product = data.productsResearch.find((product) => product.productId === filter.productId);
+		return Promise.resolve(product) as any;
+	}
+}
```
[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/examples/federation/src/backend/schema/product-research.ts:11]: That's not an `async` function though, is it?

[the remark this answers is not in the document]
Comment (thekevinbrown) [reply src/examples/federation/src/backend/schema/product-research.ts:11]: You said "The current implementation of findOne could be optimised by directly returning the result without wrapping it in Promise.resolve, as it is redundant in an async function.", however you also said "you're correct that the findOne function in the JsonDataProvider class isn't explicitly declared as async." You recommended not using `Promise.resolve` here. How do you recommend we return a promise without wrapping it in `Promise.resolve` here?

File: src/examples/federation/src/backend/schema/product.ts [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+import { ProductDimension } from './product-dimension';
+import { ProductResearch } from './product-research';
+import { User } from './user';
+import { data } from '../data';
+
+class JsonDataProvider extends BaseDataProvider<Product> {
+	findOne(filter: Filter<Product>): Promise<Product> {
+		const product = data.products.find((product) => product.id === filter.id);
+		return Promise.resolve(product) as any;
+	}
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/examples/federation/src/backend/schema/product.ts]: It is hardcoded data so will always be there

File: src/packages/cli/bin/gw-print-schema.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+printSchema()
+	.then(() => {
+		process.exit(0);
+	})
+	.catch((e) => {
+		console.error(
+			`\n\nPrinting Schema Error: \n\n${e?.message ?? 'An error occurred while printing schema.'}\n\nPlease resolve the above error to print.`
+		);
+		process.exit(1);
+	});
```
[the remark this answers is not in the document]
Comment (steve-keep) [reply src/packages/cli/bin/gw-print-schema.js:43]: We just catch the error and exit with 1

File: src/examples/federation/src/backend/index.ts [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+	await hapi.register({
+		plugin: hapiApollo,
+		options: {
+			apolloServer: graphweaver.server,
+			path: '/',
+		},
+	});
+
+	await hapi.start();
```
Comment (thekevinbrown) [inline src/examples/federation/src/backend/index.ts:44]: Why do we need both the Graphweaver server and Hapi here? Could we start it with `graphweaver start` using `sls offline` like we normally do?

File: src/packages/cli/bin/gw-print-schema.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+		);
+		process.exit(0);
+	}
+
+	const sdl = utils.printSchemaWithDirectives(graphweaver.schema);
+
+	const args = process.argv;
+
+	if (args.includes('--output') || args.includes('-o')) {
+		const outputIndex = args.indexOf('--output') + 1 || args.indexOf('-o') + 1;
```
Comment (thekevinbrown) [inline src/packages/cli/bin/gw-print-schema.js:23]: We should be able to just pass this arg in already parsed and pulled out by `yargs` so we don't have to dig around in `process.argv` ourselves.

File: src/packages/end-to-end/scripts/import-federation.ts [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+		const packageJson = JSON.parse(await fs.promises.readFile(packageJsonPath, 'utf-8'));
+
+		for (const key of Object.keys(packageJson.dependencies ?? {})) {
+			if (key.startsWith('@exogee')) {
+				packageJson.dependencies[key] = `file:../local_modules/${key}`;
+			} else if (key === 'graphweaver') {
+				packageJson.dependencies[key] = `file:../local_modules/graphweaver`;
+			} else if (key === 'vite-plugin-graphweaver') {
+				packageJson.dependencies[key] = `file:../local_modules/vite-plugin-graphweaver`;
+			}
```
Comment (thekevinbrown) [inline src/packages/end-to-end/scripts/import-federation.ts]: We could simplify this with:

```typescript
if (key.startsWith('@exogee') || key === 'graphweaver' || key === 'vite-plugin-graphweaver') {
	packageJson.dependencies[key] = `file:../local_modules/${key}`;
}
```

File: src/packages/end-to-end/scripts/import-federation.ts [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+		}
+
+		for (const key of Object.keys(packageJson.devDependencies ?? {})) {
+			if (key.startsWith('@exogee')) {
+				packageJson.dependencies[key] = `file:../local_modules/${key}`;
+			} else if (key === 'graphweaver') {
+				packageJson.dependencies[key] = `file:../local_modules/graphweaver`;
+			} else if (key === 'vite-plugin-graphweaver') {
+				packageJson.dependencies[key] = `file:../local_modules/vite-plugin-graphweaver`;
+			}
```
Comment (thekevinbrown) [inline src/packages/end-to-end/scripts/import-federation.ts]: Same here.

**Review by thekevinbrown — CHANGES_REQUESTED** [review]
The core of this looks great, just some small clean up around the edges before we land it, feel free to chat!

File: src/examples/federation/src/backend/index.ts (hunk 7 above)
Comment (steve-keep) [reply src/examples/federation/src/backend/index.ts:44]: not easily as I would need to bring in all the workspace into the docker.

This seemed like the easiest solution, what do you think?

File: src/packages/cli/bin/gw-print-schema.js (hunk 8 above)
Comment (steve-keep) [reply src/packages/cli/bin/gw-print-schema.js:23]: This script is started as an external bin, so I don't think I can pass it in.

File: src/packages/end-to-end/scripts/import-federation.ts (hunk 9 above)
Comment (steve-keep) [reply src/packages/end-to-end/scripts/import-federation.ts]: Deleted this file as no longer needed.

File: src/packages/end-to-end/scripts/import-federation.ts (hunk 10 above)
Comment (steve-keep) [reply src/packages/end-to-end/scripts/import-federation.ts]: Same here 😄 its deleted

File: src/examples/federation/src/backend/index.ts (hunk 7 above)
Comment (thekevinbrown) [reply src/examples/federation/src/backend/index.ts:44]: Yup, sold.

File: src/packages/cli/bin/gw-print-schema.js (hunk 8 above)
Comment (thekevinbrown) [reply src/packages/cli/bin/gw-print-schema.js:23]: Got it, within the docker container.

**Review by thekevinbrown — APPROVED** [review]

**Pull request merged**